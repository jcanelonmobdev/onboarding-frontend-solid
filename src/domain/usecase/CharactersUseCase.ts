import { API_URL } from "../../config/constants";
import { Character } from "../entity/Character";
import Info from "../entity/Info";


export interface CharactersResponse {
    info?: Info,
    results: Character[]
}

export class CharactersUseCase {
    list(): Promise<CharactersResponse> {
        return fetch(`${API_URL}/character`).then(response => response.json())
    }

    find(id: number): Promise<Character> {
        return fetch( `${API_URL}/character/${id}`).then(response => response.json())
    }
}
