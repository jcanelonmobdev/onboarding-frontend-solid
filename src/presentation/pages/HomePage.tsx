import React, { useEffect, useState } from 'react';
import { Character } from "../../domain/entity/Character";
import { CharactersUseCase } from "../../domain/usecase/CharactersUseCase";
import { CharacterCard } from "../component/CharacterCard";
const HomePage: React.FC = () => {

    const charactersUseCase = new CharactersUseCase()
    const [characters, setCharacters] = useState<Character[]>([])

    useEffect(() => {
        charactersUseCase.list().then((charactersData)  => {
            setCharacters(charactersData.results)
        }).catch(e => {
            console.error(e)
            return []
        })
    }, [])

    return (
        <div className="section">
            <div className="container">
                <div className="columns is-fullwidth is-multiline">
                    {characters.map((character: Character, i: number) => (
                        <div key={`character${i}`} className="column is-one-quarter">
                            <CharacterCard character={character}/>
                        </div>
                    ))}
                </div>
            </div>
        </div>
    )
}

export default HomePage
