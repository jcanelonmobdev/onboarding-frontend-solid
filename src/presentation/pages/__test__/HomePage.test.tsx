import { act } from "react-dom/test-utils";
import HomePage from "../HomePage";
import { render } from "react-dom";
import { CharactersResponse, CharactersUseCase } from "../../../domain/usecase/CharactersUseCase";
import { CHARACTERS_LIST } from "../../../domain/entity/__tests__/__mocks__/characters";
jest.mock("../../../domain/usecase/CharactersUseCase")


describe('HomePage', () => {
    let container: HTMLDivElement;

    beforeEach(() => {
        container = document.createElement('div')
        document.body.appendChild(container)
        mockCharactersUseCase()
    })

    const mockCharactersUseCase = () => {
        // @ts-ignore
        CharactersUseCase.mockImplementation(() => {
            return {
                list: (): Promise<CharactersResponse> => {
                    return Promise.resolve({results: CHARACTERS_LIST})
                }
            }
        })
    }


    it ('Home snapshots', async () => {
        await act(()=> {
            const element = <HomePage/>
            render(element, container)
        })
        expect(container).toMatchSnapshot()
    })

    it ('HomePage renders correct cards number', async() => {
        await act(()=> {
            const element = <HomePage/>
            render(element, container)
        })
        const cards = document.querySelectorAll('.character-card')
        expect(cards.length).toBe(CHARACTERS_LIST.length)
    })
})
