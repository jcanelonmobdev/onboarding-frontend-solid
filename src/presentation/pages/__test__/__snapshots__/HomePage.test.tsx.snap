// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HomePage Home snapshots 1`] = `
<div>
  <div
    class="section"
  >
    <div
      class="container"
    >
      <div
        class="columns is-fullwidth is-multiline"
      >
        <div
          class="column is-one-quarter"
        >
          <div
            class="card"
          >
            <div
              class="card-image"
            >
              <figure
                class="image is-4by3"
              >
                <img
                  alt="Placeholder image"
                  src="https://rickandmortyapi.com/api/character/avatar/1.jpeg"
                />
              </figure>
            </div>
            <div
              class="card-content"
            >
              <div
                class="media"
              >
                <div
                  class="media-left"
                >
                  <figure
                    class="image is-48x48"
                  >
                    <img
                      alt="Placeholder image"
                      src="https://rickandmortyapi.com/api/character/avatar/1.jpeg"
                    />
                  </figure>
                </div>
                <div
                  class="media-content"
                >
                  <p
                    class="title is-4"
                  >
                    Rick Sanchez
                  </p>
                  <p
                    class="subtitle is-6"
                  >
                    Earth (C-137)
                  </p>
                </div>
              </div>
              <div
                class="content"
              >
                Human
                <br />
                <time
                  datetime="2016-1-1"
                >
                  2017-11-04T18:48:46.250Z
                </time>
              </div>
            </div>
          </div>
        </div>
        <div
          class="column is-one-quarter"
        >
          <div
            class="card"
          >
            <div
              class="card-image"
            >
              <figure
                class="image is-4by3"
              >
                <img
                  alt="Placeholder image"
                  src="https://rickandmortyapi.com/api/character/avatar/2.jpeg"
                />
              </figure>
            </div>
            <div
              class="card-content"
            >
              <div
                class="media"
              >
                <div
                  class="media-left"
                >
                  <figure
                    class="image is-48x48"
                  >
                    <img
                      alt="Placeholder image"
                      src="https://rickandmortyapi.com/api/character/avatar/2.jpeg"
                    />
                  </figure>
                </div>
                <div
                  class="media-content"
                >
                  <p
                    class="title is-4"
                  >
                    Morty Smith
                  </p>
                  <p
                    class="subtitle is-6"
                  >
                    Earth (C-137)
                  </p>
                </div>
              </div>
              <div
                class="content"
              >
                Human
                <br />
                <time
                  datetime="2016-1-1"
                >
                  2017-11-04T18:50:21.651Z
                </time>
              </div>
            </div>
          </div>
        </div>
        <div
          class="column is-one-quarter"
        >
          <div
            class="card"
          >
            <div
              class="card-image"
            >
              <figure
                class="image is-4by3"
              >
                <img
                  alt="Placeholder image"
                  src="https://rickandmortyapi.com/api/character/avatar/3.jpeg"
                />
              </figure>
            </div>
            <div
              class="card-content"
            >
              <div
                class="media"
              >
                <div
                  class="media-left"
                >
                  <figure
                    class="image is-48x48"
                  >
                    <img
                      alt="Placeholder image"
                      src="https://rickandmortyapi.com/api/character/avatar/3.jpeg"
                    />
                  </figure>
                </div>
                <div
                  class="media-content"
                >
                  <p
                    class="title is-4"
                  >
                    Summer Smith
                  </p>
                  <p
                    class="subtitle is-6"
                  >
                    Earth (Replacement Dimension)
                  </p>
                </div>
              </div>
              <div
                class="content"
              >
                Human
                <br />
                <time
                  datetime="2016-1-1"
                >
                  2017-11-04T19:09:56.428Z
                </time>
              </div>
            </div>
          </div>
        </div>
        <div
          class="column is-one-quarter"
        >
          <div
            class="card"
          >
            <div
              class="card-image"
            >
              <figure
                class="image is-4by3"
              >
                <img
                  alt="Placeholder image"
                  src="https://rickandmortyapi.com/api/character/avatar/4.jpeg"
                />
              </figure>
            </div>
            <div
              class="card-content"
            >
              <div
                class="media"
              >
                <div
                  class="media-left"
                >
                  <figure
                    class="image is-48x48"
                  >
                    <img
                      alt="Placeholder image"
                      src="https://rickandmortyapi.com/api/character/avatar/4.jpeg"
                    />
                  </figure>
                </div>
                <div
                  class="media-content"
                >
                  <p
                    class="title is-4"
                  >
                    Beth Smith
                  </p>
                  <p
                    class="subtitle is-6"
                  >
                    Earth (Replacement Dimension)
                  </p>
                </div>
              </div>
              <div
                class="content"
              >
                Human
                <br />
                <time
                  datetime="2016-1-1"
                >
                  2017-11-04T19:22:43.665Z
                </time>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
